File: solutions/csharp/yacht/1/Yacht.cs
public enum YachtCategory
{
    Ones = 1,
    Twos = 2,
    Threes = 3,
    Fours = 4,
    Fives = 5,
    Sixes = 6,
    FullHouse = 7,
    FourOfAKind = 8,
    LittleStraight = 9,
    BigStraight = 10,
    Choice = 11,
    Yacht = 12,
}

public static class YachtGame
{
    private static readonly int[] LittleStraight = [1, 2, 3, 4, 5];
    private static readonly int[] BigStraight = [2, 3, 4, 5, 6];

    public static int Score(int[] dice, YachtCategory category)
    {
        var counts = CountsByFace(dice);

        return category switch
        {
            >= YachtCategory.Ones and <= YachtCategory.Sixes => SumFace(dice, (int)category),
            YachtCategory.FullHouse => IsFullHouse(counts) ? dice.Sum() : 0,
            YachtCategory.FourOfAKind => FourOfAKindScore(counts),
            YachtCategory.LittleStraight => IsStraight(dice, LittleStraight) ? 30 : 0,
            YachtCategory.BigStraight => IsStraight(dice, BigStraight) ? 30 : 0,
            YachtCategory.Choice => dice.Sum(),
            YachtCategory.Yacht => counts.Count == 1 ? 50 : 0,
            _ => 0
        };
    }

    private static int SumFace(int[] dice, int face) =>
        dice.Where(die => die == face).Sum();

    private static bool IsFullHouse(Dictionary<int, int> counts) =>
        counts.Values.Order().SequenceEqual([2, 3]);

    private static int FourOfAKindScore(Dictionary<int, int> counts) =>
        counts.Where(group => group.Value >= 4)
            .Select(group => group.Key * 4)
            .FirstOrDefault();

    private static bool IsStraight(int[] dice, int[] expected) =>
        dice.Order().SequenceEqual(expected);

    private static Dictionary<int, int> CountsByFace(int[] dice) =>
        dice.GroupBy(die => die)
            .ToDictionary(group => group.Key, group => group.Count());
}

File: solutions/scala/yacht/2/src/main/scala/Yacht.scala
object Yacht {
  private type Die = Int
  private type Dice = List[Die]

  private val LittleStraight: Dice = List(1, 2, 3, 4, 5)
  private val BigStraight: Dice = List(2, 3, 4, 5, 6)

  def score(dice: Dice, category: String): Int =
    val frequencies = frequenciesFor(dice)

    category match {
      case "ones" => sumOf(dice, 1)
      case "twos" => sumOf(dice, 2)
      case "threes" => sumOf(dice, 3)
      case "fours" => sumOf(dice, 4)
      case "fives" => sumOf(dice, 5)
      case "sixes" => sumOf(dice, 6)
      case "full house" => fullHouseScore(dice, frequencies)
      case "four of a kind" => fourOfAKindScore(frequencies)
      case "little straight" => straightScore(frequencies, LittleStraight)
      case "big straight" => straightScore(frequencies, BigStraight)
      case "choice" => dice.sum
      case "yacht" => if frequencies.contains(5) then 50 else 0
      case _ => 0
    }

  private def frequenciesFor(dice: Dice): Array[Int] =
    val frequencies = Array.fill(7)(0)
    dice.foreach(die => frequencies(die) += 1)
    frequencies

  private def sumOf(dice: Dice, number: Die): Int =
    dice.filter(_ == number).sum

  private def fullHouseScore(dice: Dice, frequencies: Array[Int]): Int =
    if frequencies.contains(2) && frequencies.contains(3) then dice.sum else 0

  private def fourOfAKindScore(frequencies: Array[Int]): Int =
    (1 to 6)
      .find(die => frequencies(die) >= 4)
      .fold(0)(_ * 4)

  private def straightScore(frequencies: Array[Int], straight: Dice): Int =
    if straight.forall(die => frequencies(die) == 1) then 30 else 0
}
